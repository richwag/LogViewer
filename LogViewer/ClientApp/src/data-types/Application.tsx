export interface Application {
    ApplicationId: string;
    ApplicationName: string;
}

