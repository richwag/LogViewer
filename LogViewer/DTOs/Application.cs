

namespace LogViewer.LogViewer.DTOs
{
    public partial class Application
    {
        public int ApplicationId { get; set; }
        public string ApplicationName { get; set; } = null!;
    }
}
