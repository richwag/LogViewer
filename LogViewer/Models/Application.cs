using System.ComponentModel.DataAnnotations;

namespace LogViewer.LogViewer.Models
{
    public partial class Application
    {
        public Application()
        {
            ApplicationLogs = new HashSet<LogEntry>();
        }

        [Key]
        public int ApplicationId { get; set; }
        public string ApplicationName { get; set; } = null!;

        public virtual ICollection<LogEntry> ApplicationLogs { get; set; }
    }
}
