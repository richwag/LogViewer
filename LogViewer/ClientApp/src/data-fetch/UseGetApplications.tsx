import { useEffect, useState } from "react";
import { config } from "../config";
import { Application } from "../data-types/Application";

const pleaseChoose: Application = {
    ApplicationId: "-1",
    ApplicationName: "Please Choose",
};

function useGetApplications(
    environment: string | null
): [Array<Application>, boolean, string | null] {
    const [apps, setApps] = useState<Array<Application>>([]);
    const [busy, setBusy] = useState(false);
    const [error, setError] = useState<string | null>(null);

    useEffect(() => {
        const fetchData = async () => {
            try {
                setError(null);
                setBusy(true);

                const response = await fetch(
                    `${config.server_prefix_url}ErrorManagement/ApplicationInfos?environment=${environment}`,
                    {
                        mode: "cors",
                        credentials: "include",
                    }
                );

                setBusy(false);

                if (response.ok) {
                    var tmpApps: Array<Application> = await response.json();
                    tmpApps.sort((a, b) => {
                        return a.ApplicationName.localeCompare(
                            b.ApplicationName
                        );
                    });
                    tmpApps.splice(0, 0, pleaseChoose);
                    setApps(tmpApps);
                } else {
                    throw Error("Can't retrieve applications");
                }
            } catch (e) {
                setError("Can't retrieve applications");
                setBusy(false);
            }
        };

        if (environment !== null && environment.length) {
            fetchData();
        }
    }, [environment]);

    return [apps, busy, error];
}

export { useGetApplications };
