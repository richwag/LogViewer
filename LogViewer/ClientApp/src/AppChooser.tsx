import { Spinner } from "react-bootstrap";
import { Application } from "./Application";
import { useGetApplications } from "./UseGetApplications";
import { FieldError } from "./FieldError";
import { useEnvironment } from "./EnvironmentProvider";

// AppChooser
export function AppChooser(props: { setApp: Function }) {
    const environment = useEnvironment();
    const [apps, busy, error] = useGetApplications(environment);

    function appChange(applicationId: string) {
        if (applicationId !== "-1") {
            props.setApp(
                apps.find(
                    (a: Application) =>
                        a.ApplicationId.toString() === applicationId
                )
            );
        } else {
            props.setApp(undefined);
        }
    }

    return (
        <>
            <label>Application</label>
            <div>
                {!busy && !error && (
                    <select
                        className="form-control"
                        onChange={(e) => appChange(e.target.value)}
                    >
                        {apps.map((a: Application) => (
                            <option
                                key={a.ApplicationId}
                                value={a.ApplicationId}
                            >
                                {a.ApplicationName}
                            </option>
                        ))}
                    </select>
                )}
                {!!busy && (
                    <Spinner
                        as="span"
                        animation="border"
                        size="sm"
                        role="status"
                        aria-hidden="true"
                    />
                )}
                {error && <FieldError errorMessage={error} show={!!error} />}
            </div>
        </>
    );
}
