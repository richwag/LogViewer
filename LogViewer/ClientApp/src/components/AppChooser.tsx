import { Spinner } from "react-bootstrap";
import { useGetApplications } from "../data-fetch/UseGetApplications";
import { Application } from "../data-types/Application";
import { useEnvironment } from "../EnvironmentProvider";
import { FieldError } from "./FieldError";

// AppChooser
export function AppChooser(props: { setApp: Function }) {
    const environment = useEnvironment();
    const [apps, busy, error] = useGetApplications(environment);

    function appChange(applicationId: string) {
        if (applicationId !== "-1") {
            props.setApp(
                apps.find(
                    (a: Application) =>
                        a.ApplicationId.toString() === applicationId
                )
            );
        } else {
            props.setApp(undefined);
        }
    }

    return (
        <>
            <label>Application</label>
            <div>
                {!busy && !error && (
                    <select
                        title="Choose an application"
                        className="form-control"
                        onChange={(e) => appChange(e.target.value)}
                    >
                        {apps.map((a: Application) => (
                            <option
                                key={a.ApplicationId}
                                value={a.ApplicationId}
                            >
                                {a.ApplicationName}
                            </option>
                        ))}
                    </select>
                )}
                {!!busy && (
                    <Spinner
                        as="span"
                        animation="border"
                        size="sm"
                        role="status"
                        aria-hidden="true"
                    />
                )}
                {error && <FieldError errorMessage={error} show={!!error} />}
            </div>
        </>
    );
}
