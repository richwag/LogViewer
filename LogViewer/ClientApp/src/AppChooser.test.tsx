import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { AppChooser } from "./AppChooser";
import { Application } from "./Application";
import userEvent from "@testing-library/user-event";
import * as hooks from "./UseGetApplications";

const mockApps: Array<Application> = [
    {
        ApplicationName: "application",
        ApplicationId: "1",
    },
];

beforeEach(() => {});

afterEach(() => {});

describe("AppChooser", () => {
    it("renders", async () => {
        const setApp = jest.fn();

        jest.spyOn(hooks, "useGetApplications").mockImplementation(() => [
            mockApps,
            false,
            null,
        ]);

        let result = render(<AppChooser setApp={setApp} />);

        expect(result.container).toMatchSnapshot();
    });

    it("calls setApp when app is selected", () => {
        const setApp = jest.fn();

        jest.spyOn(hooks, "useGetApplications").mockImplementation(() => [
            mockApps,
            false,
            null,
        ]);

        let result = render(<AppChooser setApp={setApp} />);

        let select = result.container.querySelector("select");

        expect(select).toBeTruthy();

        select && userEvent.selectOptions(select, "1");

        expect(setApp).toBeCalledWith(mockApps[0]);
    });

    it("shows error when get applications fails", async () => {
        const setApp = jest.fn();

        jest.spyOn(hooks, "useGetApplications").mockImplementation(() => [
            [],
            false,
            "Error!!!",
        ]);

        let result = render(<AppChooser setApp={setApp} />);

        let select = result.container.querySelector("select");
        expect(select).toBeFalsy();

        expect(await screen.findByText("Error!!!")).toBeTruthy();
    });

    it("shows spinner when loading", async () => {
        const setApp = jest.fn();

        jest.spyOn(hooks, "useGetApplications").mockImplementation(() => [
            [],
            true,
            null,
        ]);

        let result = render(<AppChooser setApp={setApp} />);

        let select = result.container.querySelector("select");
        expect(select).toBeFalsy();

        expect(screen.queryAllByRole("status")).toBeTruthy();
    });
});
