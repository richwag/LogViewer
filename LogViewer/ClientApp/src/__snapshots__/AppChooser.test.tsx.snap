// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppChooser renders 1`] = `
<div>
  <label>
    Application
  </label>
  <div>
    <select
      class="form-control"
    >
      <option
        value="1"
      >
        application
      </option>
    </select>
  </div>
</div>
`;
